cdap-runtime-ext-dataproc/src/main/java/co/cask/cdap/runtime/spi/provisioner/dataproc/DataProcConf.java

```diff
@@ -213,15 +217,23 @@ private static DataProcConf create(Map<String, String> properties, @Nullable SSH
     long pollDeleteDelay = getLong(properties, "pollDeleteDelay", 30);
     long pollInterval = getLong(properties, "pollInterval", 2);
 
+    boolean preferExternalIP = Boolean.parseBoolean(properties.get("preferExternalIP"));
+
     return new DataProcConf(accountKey, region, zone, projectId, network,
                             masterNumNodes, masterCPUs, masterMemoryGB, masterDiskGB,
                             workerNumNodes, workerCPUs, workerMemoryGB, workerDiskGB,
-                            pollCreateDelay, pollCreateJitter, pollDeleteDelay, pollInterval, publicKey);
+                            pollCreateDelay, pollCreateJitter, pollDeleteDelay, pollInterval,
+                            preferExternalIP, publicKey);
   }
 
-  private static String getString(Map<String, String> properties, String key, String defaultVal) {
+  // the UI never sends nulls, it only sends empty strings.
```

Comment:
"UI never sends nulls"
I think this is true for some widget types, such as text widget type.
But its not true for all widget types, such as selector widgets.

Reply:
selectors don't have a way to not select anything (at least in the provisioner page)

Reply:
oh ok; my bad. I must have misunderstood something.